### PR TITLE
Add total_length option to pad_packed_sequence

### LINKED ISSUE #6299: [feature request] make pad_packed_sequence work in DataParallel
Currently users can not do `pack -> RNN -> unpack` in a module wrapped in `DataParallel` because the unpack operation (`pad_packed_sequence`) will only pad up to the longest input it sees, i.e., the longest on that particular device. Then the code breaks when it tries to gather the results into a single tensor afterwards. I propose to add a `total_length` option to `pad_packed_sequence` so that it can beyond the longest input. Then the following pattern should work with DataParallel:
```python
def forward(self, input, input_lengths):
    total_length = input.size(-1)
    packed_input = nn.utils.rnn.pack_padded_sequence(input, input_lengths)
    packed_output = self.lstm(packed_input)
    output, _ = nn.utils.rnn.pad_packed_sequence(packed_output, total_length=total_length)
    return output
```

The current suggested workaround is to manually add the padding. However, it becomes tricky when one wants to do ONNX export.

Relevant post: https://discuss.pytorch.org/t/question-about-packed-rnn-with-dataparallel/2738

cc @goldsborough

### PR DESCRIPTION
**Pull request by ssnl**
Add example on how to use pack->rnn->unpack with DataParallel.
Fix incorrect comment on hooks with DataParallel.

Fixes #6299

### CODE REVIEW

**Comment by SsnL** [comment]
Requesting a review from @apaszke

File: test/test_nn.py [hunk 1]
```diff
@@ -128,6 +128,26 @@ def test_cuda_mask(self):
...
+        padded, lengths = self._padded_sequence(torch.FloatTensor)
+        max_length = max(lengths)
+        packed = rnn_utils.pack_padded_sequence(padded, lengths)
+        # test ValueError if total_length < max_length
+        for total_length in (-1, 0, max_length - 1):
+            for batch_first in (True, False):
+                def err_fn():
+                    rnn_utils.pad_packed_sequence(packed, batch_first=batch_first,
+                                                  total_length=total_length)
+            self.assertRaises(ValueError, err_fn)
```
Comment (apaszke) [inline test/test_nn.py]: Can you please test for the error message too (just to make sure it is that error)

File: test/test_nn.py [hunk 2]
```diff
@@ -128,6 +128,26 @@ def test_cuda_mask(self):
...
+        # test that pad_packed_sequence returns results of correct length
+        for total_length_delta in (0, 1, 8):
+            for batch_first in (True, False):
+                total_length = max_length + total_length_delta
+                unpacked, lengths_out = rnn_utils.pad_packed_sequence(packed, batch_first=batch_first,
+                                                                      total_length=total_length)
+                self.assertEqual(lengths, lengths_out)
+                self.assertEqual(unpacked.size(1 if batch_first else 0), total_length)
```
Comment (apaszke) [inline test/test_nn.py]: Can you please also check that the extra padding is 0 and that the data is still laid out correctly (you can use `pad_packed_sequence` without `total_length`, `cat` and check for equality)?

File: torch/nn/parallel/data_parallel.py [hunk 3]
```diff
@@ -52,8 +52,51 @@ class DataParallel(Module):
...
+                    packed_input = pack_padded_sequence(padded_input, input_lengths,
+                                                        batch_first=True)
+                    packed_output, _ = self.my_lstm(packed_input)
+                    output, _ = pad_packed_sequence(packed_output, batch_first=True,
+                                                    total_length=total_length)
+                    return output
+
+
+            m = MyModule().cuda()
+            dp_m = nn.DataParallel(m)
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: I think that's a bit too much for the API docs. We might want to put it in some kind of FAQ and put a reference to it here

File: torch/nn/utils/rnn.py [hunk 4]
```diff
@@ -145,7 +146,7 @@ def _symbolic_pad_packed_sequence(g, input, batch_first=False, padding_value=0.0
 
 
 @torch.onnx.symbolic_override_packed_sequence_based(_symbolic_pad_packed_sequence)
-def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
+def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0, total_length=None):
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: The symbolics should probably raise if they see `total_length` being specified

File: torch/onnx/__init__.py [hunk 5]
```diff
@@ -124,7 +124,8 @@ def symbolic_override_first_arg_based(symbolic_fn):
 
     def might_trace(args):
         import torch
-        return torch._C._jit_is_tracing(args[0])
+        first_arg = args[0]
+        return torch.is_tensor(first_arg) and torch._C._jit_is_tracing(first_arg)
```
Comment (apaszke) [inline torch/onnx/__init__.py]: I don't think we should silently skip if it's not a tensor, because it's likely an error

File: torch/onnx/__init__.py [hunk 6]
```diff
@@ -140,6 +141,11 @@ def symbolic_override_packed_sequence_based(symbolic_fn):
 
     def might_trace(args):
         import torch
-        return torch._C._jit_is_tracing(args[0][0])
+        first_arg = args[0]
+        if not isinstance(first_arg, torch.nn.utils.rnn.PackedSequence):
+            raise ValueError('pad_packed_sequence expects sequence to be a '
+                             'PackedSequence, but got an object of type {}'
+                             .format(type(first_arg)))
+        return torch.is_tensor(first_arg[0]) and torch._C._jit_is_tracing(first_arg[0])
```
Comment (apaszke) [inline torch/onnx/__init__.py]: This error message looks nice, let's add something like it above. Also, I think the `is_tensor` check is redundant here, as `.data` in `PackedSequence` is always a tensor

**Review by apaszke — COMMENTED** [review]
LGTM! A few minor requests.

File: test/test_nn.py (hunk 2 above)
Comment (SsnL) [reply test/test_nn.py]: added one below :)

File: torch/nn/utils/rnn.py [hunk 7]
```diff
@@ -136,7 +137,10 @@ def pack_padded_sequence(input, lengths, batch_first=False):
     return PackedSequence(data, batch_sizes)
 
 
-def _symbolic_pad_packed_sequence(g, input, batch_first=False, padding_value=0.0):
+def _symbolic_pad_packed_sequence(g, input, batch_first=False, padding_value=0.0, total_length=None):
+    if total_length is not None:
+        raise ValueError("_symbolic_pad_packed_sequence only supports "
+                         "_symbolic_pad_packed_sequence=None")
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: you probably meant `total_length=None` 😄

**Review by apaszke — APPROVED** [review]
LGTM! There's one minor typo left, and feel free to merge once that's fixed

File: torch/nn/utils/rnn.py (hunk 7 above)
Comment (SsnL) [reply torch/nn/utils/rnn.py]: haha yeah, let me update that!

File: torch/onnx/__init__.py (hunk 6 above)
Comment (fmassa) [reply torch/onnx/__init__.py]: Btw, should we be using `isinstance(first_arg[0], torch.Tensor)` ?

Comment (apaszke) [reply torch/onnx/__init__.py]: We should. I forgot about it
